### PR TITLE
INGM-266 Improve load_FWs script using canopen protocol

### PR DESCRIPTION
**Pull request by pzinemanas**
### Description

This PR improves the script to load FW using canopen.

Fixes # INGM-266

### Type of change

- [x] Add attempts when loading FW.
- [x] Verify the drive responds after flashing.

### Tests
- [x] Run the script locally and in the CI setup.

### CODE REVIEW

**Review by GullinBustin — APPROVED** [review]

File: tests/load_FWs.py [hunk 1]
```diff
@@ -25,35 +25,98 @@ def setup_command():
...
+    BL_NUM_OF_REATTEMPTS = 2
+
+    # Timings, in seconds
+    SLEEP_TIME_AFTER_ATTEMP = 5.0
+    SLEEP_TIME_AFTER_BL = 5.0
+    TIMEOUT_NEW_FW_DETECT = 30.0
+    SLEEP_TIME_NEW_FW_DETECT = 5.0
+
+    for attempt in range(BL_NUM_OF_REATTEMPTS):
+        logger.info("CAN boot attempt {} of {}".format(attempt + 1, BL_NUM_OF_REATTEMPTS))
```
Comment (MartinALL91) [inline tests/load_FWs.py:38]: Use f-string

File: tests/load_FWs.py [hunk 2]
```diff
@@ -25,35 +25,98 @@ def setup_command():
...
+        try:
```
Comment (MartinALL91) [inline tests/load_FWs.py:39]: I think this try should only wrap up the `load_firmware_canopen` function.

File: tests/load_FWs.py [hunk 3]
```diff
@@ -25,35 +25,98 @@ def setup_command():
...
+            time.sleep(SLEEP_TIME_AFTER_ATTEMP)
+            break
+
+        except ILFirmwareLoadError as e:
+            # Disconnect CAN network
+            mc.communication.disconnect()
+
+            if str(e) != "Could not recover drive":
+                # TODO Fix canopen FW loader and remove this if
+                err_str = "CAN boot error: {}".format(e)
```
Comment (MartinALL91) [inline tests/load_FWs.py:70]: Use f-string

File: tests/load_FWs.py [hunk 4]
```diff
@@ -25,35 +25,98 @@ def setup_command():
...
+                logger.error(err_str)
+                time.sleep(SLEEP_TIME_AFTER_ATTEMP)
```
Comment (MartinALL91) [inline tests/load_FWs.py:72]: I don't think this sleep is necessary.

File: tests/load_FWs.py [hunk 5]
```diff
@@ -25,35 +25,98 @@ def setup_command():
...
+                raise e
+            else:
+                err_str = None
+                logger.warning(f"Exception '{e}' has raised, but it is ignored.")
+                break
+
+    if err_str is None:
```
Comment (MartinALL91) [inline tests/load_FWs.py:79]: I think the `err_str` is not needed, if you reach this line it means the firmware was loaded successfully

File: tests/load_FWs.py [hunk 6]
```diff
@@ -25,35 +25,98 @@ def setup_command():
...
+            "FW updated. %s, node: %d, baudrate: %d, channel: %d",
+            drive_conf["device"],
+            drive_conf["node_id"],
+            drive_conf["baudrate"],
+            drive_conf["channel"],
+        )
+    else:
+        raise Exception(err_str)
+
+    logger.info("Waiting {} seconds for trying to connect".format(SLEEP_TIME_AFTER_BL))
```
Comment (MartinALL91) [inline tests/load_FWs.py:90]: Use f-string

File: tests/load_FWs.py [hunk 7]
```diff
@@ -25,35 +25,98 @@ def setup_command():
...
+            # When cannot connect
+            pass
+
+        mc.communication.disconnect()
+
+        if not detected:
+            time.sleep(SLEEP_TIME_NEW_FW_DETECT)
+
+    if not detected:
+        raise Exception("New FW not detected")
```
Comment (MartinALL91) [inline tests/load_FWs.py:118]: Maybe we should use the `ILFirmwareLoadError` exception.

File: tests/load_FWs.py [hunk 8]
```diff
@@ -25,35 +25,98 @@ def setup_command():
...
+            detected = True
+            logger.info("New FW detected after: {:.1f} s".format(time.perf_counter() - ini_time))
+        except Exception as e:
+            # When cannot connect
+            pass
+
+        mc.communication.disconnect()
+
+        if not detected:
+            time.sleep(SLEEP_TIME_NEW_FW_DETECT)
```
Comment (MartinALL91) [inline tests/load_FWs.py:115]: Maybe we could move this sleep to the top of the while loop and remove the if statement.

**Review by MartinALL91 — APPROVED** [review]

File: tests/load_FWs.py (hunk 7 above)
Comment (pzinemanas) [reply tests/load_FWs.py:110]: I thought that script-level exceptions should be general `Exceptions` and let `ILFirmwareLoadError` for the IL-related exceptions.

File: tests/load_FWs.py [hunk 9]
```diff
@@ -25,35 +25,89 @@ def setup_command():
...
+
+            mc.communication.disconnect()
+
+            # Reaching this means that FW was correctly flashed
+            time.sleep(SLEEP_TIME_AFTER_ATTEMP)
+            break
+
+        except ILFirmwareLoadError as e:
+            # Disconnect CAN network
+            mc.communication.disconnect()
```
Comment (MartinALL91) [inline tests/load_FWs.py:65]: You can use the `finally` clause for this command and remove it from here and from line 57.

File: tests/load_FWs.py [hunk 10]
```diff
@@ -25,35 +25,80 @@ def setup_command():
...
+                channel=drive_conf["channel"],
+            )
+            # Reaching this point means we are connected
+            detected = True
+            logger.info("New FW detected after: {:.1f} s".format(time.perf_counter() - ini_time))
+            mc.communication.disconnect()
+        except Exception as e:
+            # When cannot connect
+            time.sleep(SLEEP_TIME_NEW_FW_DETECT)
+            pass
```
Comment (MartinALL91) [inline tests/load_FWs.py:98]: I think this line is not necessary

**Review by WoodElv — APPROVED** [review]

**Pull request merged**